File: python/aurigraph_sdk/namespaces/channels.py
"""Namespace for /api/v11/channels endpoints -- channel management."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any
from urllib.parse import quote

from aurigraph_sdk.models import Channel

if TYPE_CHECKING:
    from aurigraph_sdk.client import AurigraphClient


class ChannelsApi:
    """Channel management operations."""

    def __init__(self, client: AurigraphClient) -> None:
        self._client = client

    def list(self) -> list[Channel]:
        """List all channels.

        Returns:
            List of channels. Handles both array and wrapped responses.
        """
        data = self._client._get("/channels")
        items: list[dict[str, Any]]
        if isinstance(data, list):  # type: ignore[arg-type]
            items = data  # type: ignore[assignment]
        elif isinstance(data, dict):
            items = data.get("channels", data.get("items", []))
        else:
            items = []
        return [Channel.model_validate(ch) for ch in items]

    def get(self, channel_id: str) -> Channel:
        """Get a specific channel by ID.

        Args:
            channel_id: The channel identifier.

        Returns:
            Channel information.
        """
        enc = quote(channel_id, safe="")
        data = self._client._get(f"/channels/{enc}")
        return Channel.model_validate(data)

File: rust/src/namespaces/channels.rs
//! Channel management namespace for `/api/v11/channels/*` endpoints.
//!
//! Provides channel listing and detail retrieval.
//!
//! # Example
//!
//! ```rust,no_run
//! # async fn example(client: &aurigraph_sdk::AurigraphClient) -> aurigraph_sdk::Result<()> {
//! let channels = client.channels().list().await?;
//! println!("Channels: {}", channels);
//!
//! let channel = client.channels().get("ch-1").await?;
//! println!("Channel: {}", channel);
//! # Ok(())
//! # }
//! ```

use crate::client::AurigraphClient;
use crate::error::Result;

/// Channels API namespace.
///
/// Obtained via [`AurigraphClient::channels()`].
#[derive(Debug)]
pub struct ChannelsApi<'a> {
    client: &'a AurigraphClient,
}

impl<'a> ChannelsApi<'a> {
    pub(crate) fn new(client: &'a AurigraphClient) -> Self {
        Self { client }
    }

    /// List all channels on the network.
    ///
    /// Calls `GET /api/v11/channels`.
    pub async fn list(&self) -> Result<serde_json::Value> {
        self.client.get("/channels").await
    }

    /// Get a specific channel by ID.
    ///
    /// Calls `GET /api/v11/channels/{channel_id}`.
    pub async fn get(&self, channel_id: &str) -> Result<serde_json::Value> {
        self.client
            .get(&format!("/channels/{}", channel_id))
            .await
    }
}
